Add kubestash docs (#1)
kubestash/docs: docs/concepts/crds/addon/index.md

---
title: Addon Overview
menu:
  docs_{{ .version }}:
    identifier: addon-overview
    name: Addon
    parent: crds
    weight: 30
product_name: kubestash
menu_name: docs_{{ .version }}
section_menu_id: concepts
---

> New to KubeStash? Please start [here](/docs/concepts/README.md).

# Addon

## What is Addon

> An `Addon` is a Kubernetes `CustomResourceDefinition`(CRD) which specifies the backup and restore capabilities for a particular resource. For example, MySQL addon specifies the backup and restore capabilities of MySQL database where Postgres addon specifies backup and restore capabilities for PostgreSQL database. An Addon CR defines the backup and restore tasks that can be performed by this addon.

Reviewer: ```suggestion
An `Addon` is a Kubernetes `CustomResourceDefinition`(CRD) which specifies the backup and restore capabilities for a specific type of target. For example, MySQL addon specifies the backup and restore capabilities of MySQL database where Postgres addon specifies backup and restore capabilities for PostgreSQL database. An Addon CR defines these capabilities as tasks for backup and restore.
```

        - name: kubestash-tmp-volume
          source:
            emptyDir: {}
          usage: Holds temporary files and restic cache.
```
This `Addon` has three backup tasks and two restore tasks for a PostgreSQL database. Each task refers to a `Function`.

Here, we are going to describe the various sections of a `Addon` crd.

## Addon `Spec`
A `Addon` object has the following fields in the `spec` section:
- **backupTasks :** specifies a list of backup tasks that can be performed by the addon. See [task specification](#task-specification) to learn more about backup tasks.
- **restoreTasks :** specifies a list of restore tasks that can be performed by the addon. See [task specification](#task-specification) to learn more about restore tasks.

### Task Specification
Each task contains the following fields:
- **name :** specifies the name of the task. The name of a task should indicate what this task does. For example, a name `logical-backup` indicate that this task performs a logical backup of a database.
- **function :** specifies the name of a `Function` CR that defines a container definition which will execute the backup/restore logic for a particular application.

> - **driver :** specifies the underlying tool that will be used to upload the data to the backend storage. Valid values are:
>   - **Restic :** The underlying tool is [restic](https://restic.net/).
>   - **WalG :** The underlying tool is [wal-g](https://github.com/wal-g/wal-g).
>   - **VolumeSnapshotter :** The underlying method is creating Kubernetes [VolumeSnapshot](https://kubernetes.io/docs/concepts/storage/volume-snapshots/) crd and leverages CSI driver to snapshot the PVCs.

Reviewer: ```suggestion
- **driver :** specifies the underlying tool that will be used to upload the data to the backend storage. Valid values are:
  - **Restic :** uses [restic](https://restic.net/) to upload backup data.
  - **VolumeSnapshotter :** leverages  Kubernetes [VolumeSnapshot](https://kubernetes.io/docs/concepts/storage/volume-snapshots/) crd and CSI driver to snapshot the PVCs.
```

> - **executor :** specifies the type of entity that will execute the task. For example, it can be Job(s), a sidecar container, an ephemeral container, or a Job that creates additional Jobs/Pods for executing the backup/restore logic. Valid values are:
>   - **Job :** KubeStash will create Job(s) to execute the backup/restore task.
>   - **Sidecar :** KubeStash will inject a sidecar container into the application to execute the backup/restore task.
>   - **EphemeralContainer :** KubeStash will attach an ephemeral container to the respective Pods to execute the backup/restore task.
>   - **MultiLevelJob :** KubeStash will create a Job that will create additional Jobs/Pods to execute the backup/restore task.

Reviewer: ```suggestion
- **executor :** specifies the type of entity that will execute the task. Valid values are:
  - **Job :** KubeStash will create Job(s) to execute the backup/restore task.
```

  - **name :** specifies the name of the volume.
  - **usage :** specifies the usage of the volume.
  - **source :** specifies the source of this volume or a template for volume to use.
- **volumeMounts :** specifies the mount path of the volumes specified in the `VolumeTemplate` section. These volumes will be mounted directly on the Job/Container created/injected by KubeStash operator. If the volume type is `VolumeClaimTemplate`, then KubeStash operator is responsible for creating the volume.
- **passThroughMounts :** specifies a list of volume mount for the `VolumeTemplates` that should be mounted on second level Jobs/Pods created by the first level executor Job. If the volume needs to be mounted on both first level and second level Jobs/Pods, then specify the mount in both `VolumeMounts` and `PassThroughMounts` section. If the volume type is `VolumeClaimTemplate`, then the first level job is responsible for creating the volume.

> 

Reviewer: remove it for now.
